gateway/src/node/client_handling/websocket/connection_handler/authenticated.rs

```diff
@@ -412,15 +413,19 @@ where
     /// Simultaneously listens for incoming client requests, which realistically should only be
     /// binary requests to forward sphinx packets or increase bandwidth
     /// and for sphinx packets received from the mix network that should be sent back to the client.
-    pub(crate) async fn listen_for_requests(mut self)
+    pub(crate) async fn listen_for_requests(mut self, mut shutdown: TaskClient)
     where
         S: AsyncRead + AsyncWrite + Unpin,
         St: Storage,
     {
         trace!("Started listening for ALL incoming requests...");
 
-        loop {
+        while !shutdown.is_shutdown() {
             tokio::select! {
+                biased;
```

Comment:
Note that this will mean read_websocket_message always will take priority over mix_receiver.next